### PR TITLE
Distinguish between retryable and non-retryable errors when discovering services

### LINKED ISSUE #1208: When during deployment registration the scheme of registration URI is missing, Restate waits and retries
Since this is an unrecoverable error, this should immediately respond with the error not retry.
Currently, one needs to go to the runtime log to see what is happening, or wait a long time until a response comes.

There is also a weird message about "retry limit exhausted" and then "retrying in 12 seconds" 🤔 

Affects version 0.7.1

Below is the log output:
```
2024-02-22T13:41:38.336162Z WARN restate_service_protocol::discovery
  Error when discovering deployment at address 'localhost:9080'. Retrying in 12 seconds
    error: [META0003] retry limit exhausted. Last client error: scheme missing. For more details, look at the docs with https://docs.restate.dev/references/errors#META0003
    restate.error.code: META0003
```

### PR DESCRIPTION
**Pull request by tillrohrmann**
This fixes #1208.

### CODE REVIEW

**Review by pcholakov — APPROVED** [review]
Nice! This will also help speed up CDK/CloudFormation deployments when there is an error.

File: crates/service-protocol/src/discovery.rs [hunk 1]
```diff
@@ -141,14 +142,34 @@ pub enum ServiceDiscoveryError {
...
+                StatusCode::from_u16(*status).expect("should be valid status code"),
+                StatusCode::REQUEST_TIMEOUT
+                    | StatusCode::TOO_MANY_REQUESTS
+                    | StatusCode::INTERNAL_SERVER_ERROR
+                    | StatusCode::BAD_GATEWAY
+                    | StatusCode::SERVICE_UNAVAILABLE
+                    | StatusCode::GATEWAY_TIMEOUT
+            ),
+            ServiceDiscoveryError::Client(client_error) => client_error.is_retryable(),
+            _ => false,
```
Comment (AhmedSoliman) [inline crates/service-protocol/src/discovery.rs:168]: It'd be helpful for our future selves if you write a comment here about why the default case is `false` vs why it's `true` in SdkError.

Comment (tillrohrmann) [reply crates/service-protocol/src/discovery.rs:168]: I will spell the `ServiceDiscoveryError` variants that are not retryable out to make it clear that newly added variants need to be reasoned about.

Comment (tillrohrmann) [reply crates/service-protocol/src/discovery.rs]: I've pushed a fixup to list the cases explicitly.

**Pull request closed without merging**